docs/references/architecture/adr-119-dog-mempool-gossip.md

```diff
@@ -0,0 +1,239 @@
+
+By default, the redundancy is set to 1 and impacts the gossiping of transactions as follows:
+
+
+```
+  redundancy := duplicateTxs / firstTimeTxs
+  if redundancy < r.redundancyLowerBound:
+    peer.send(Reset)
+  if redundancy > redundancyUpperBound
+    enableRouteDisabling() 
+```
+
+Redundancy is adjusted whenever a new transaction is added into the mempool.
+(TODO - revisit whether this should also be done on duplicates.)
+
+
+### Impacted areas of code
+
+#### Mempool reactor
+
+
+The bulk of the changes is constrained to the mempool reactor. 
+We introduce a new communication channel, called `Mempool Control Channel`. The channel
+is used to transmit the messages needed for the implementation of the protocol. 
+
+Additionally, the mempool reactor is extended with a 
+`gossipRouter` and a `redundancyControl` struct to keep track of the redundancy
+of transactions in the system. 
+Each node is keeps track of the routes it disabled between its peers. 
+
+```
+type gossipRouter struct {
+	mtx cmtsync.RWMutex
+	// A set of `source -> target` routes that are disabled for disseminating transactions. Source
+	// and target are node IDs.
+	disabledRoutes map[nodekey.ID]p2pIDSet
+}
+``` 
+
+
+```
+type redundancyControl struct {
+	txsPerAdjustment int64
+
+	// Pre-computed upper and lower bounds of accepted redundancy.
+	lowerBound float64
+	upperBound float64
+
+	mtx          cmtsync.RWMutex
+	firstTimeTxs int64 // number of transactions received for the first time
+	duplicates   int64 // number of duplicate transactions
+
+	// If true, do not send HaveTx messages.
+	blockHaveTx atomic.Bool
+}
+
+```
+
+
+
+#### New p2p messages
+
+The protocol introduces two new p2p messages whose protobuf definition is given below:
+
+```
+type HaveTx struct {
+	TxKey []byte 
+}
+```
+
+```
+type Reset struct {
+
+}
+```
+
+#### Monitoring
+
+This ADR introduces a set of metrics which can be used to observe the parameters of the protocol. 
+
+- `HaveTxMsgsReceived`
+- `ResetMsgsSent`
+- `DisabledRoutes`
+- `Redundancy`
+
+The impact of the protocol on operations can also be observed by looking at the following metrics that already exist:
+
+-`AlreadyReceivedTx` - the number of redundant transactions. When DOG is enabled these values should drop. 
+
+-`BytesReceived` - This metric shows the number of bytes received per message type. Without DOG, the transactions dominate the number of bytes, while, when enabled, the block parts dominate.
+
+Newly introduced metrics are: 
+
+
+- `HaveTxMsgsReceived` -  Number of HaveTx messages received (cumulative). 
+
+- `ResetMsgsSent` - Number of Reset messages sent (cumulative).
+
+- `DisabledRoutes` - Number of disabled routes.
+
+- `Redundancy` -  The current level of redundancy computed as the ratio between duplicate and unique transactions. 
+
+
+
+### Configuration parameters
+
+
+- `config.EnableDOGProtocol: bool`: Enabling or disabling the DOG protocol
```

Comment:
Ideally, on the final version we would leave just one configuration: `EnableDOGProtocol`. For the other parameters described in the spec (`TargetRedundancy`, `TargetRedundancyDeltaPercent`, and (now) `AdjustInverval`) I would try to define values based on the experiments, and then make them constants  in the code. Currently, those numbers are `TargetRedundancy = 0.5`, `TargetRedundancyDeltaPercent = 20`, and `AdjustInverval = 1000`.

Reply:
We need to decide if `EnableDOGProtocol` will be `true` by default.